lib/types.nix

```diff
@@ -245,23 +287,46 @@ rec {
           else if builtins.isInt v then builtins.toString v
           else ''<${builtins.typeOf v}>'';
       in
-      mkOptionType {
-        name = "one of ${concatMapStringsSep ", " show values}";
+      mkOptionType rec {
+        name = "enum";
+        description = "one of ${concatMapStringsSep ", " show values}";
         check = flip elem values;
         merge = mergeOneOption;
+        typeMerge = defaultTypeMerge functor;
+        functor = { type = enum; wrapped = null; payload = values; binOp = a: b: a ++ b; };
       };
 
-    either = t1: t2: mkOptionType {
-      name = "${t1.name} or ${t2.name}";
+    either = t1: t2: mkOptionType rec {
+      name = "either";
+      description = "${t1.description} or ${t2.description}";
       check = x: t1.check x || t2.check x;
       merge = mergeOneOption;
+      typeMerge = f':
+        let t1  = head   functor.wrapped;
+            t2  = elemAt functor.wrapped 1;
+            t1' = head   f'.wrapped;
+            t2' = elemAt f'.wrapped 1;
+            mt1 = t1.typeMerge t1'.functor;
+            mt2 = t2.typeMerge t2'.functor;
+        in
+           # Eithers wrapping different types are unmergeable
+           if (t1.name != t1'.name) || (t2.name != t2'.name)
+             then null
+           # Both wrapped types are unmergeable
+           else if ((mt1 == null) && (mt2 == null))
+             then null
+           else functor.type
+                  (if mt1 != null then mt1 else t1)
+                  (if mt2 != null then mt2 else t2);
```

Comment:
This is not a commutative behaviour, either both types should be mergeable, or we should not be able to merge this either type.